### ISSUE TITLE
issues with infix notation

### ISSUE THREAD

**Issue by agriggio** [pysmt/pysmt#92]
It seems that the infix notation has still some rough edges.
Some examples:
From the shell.py, using `get_env().enable_infix_notation=True`

```
x = Symbol('x', INT)
f = (x > Int(10))
```

works, but:

```
f = (x > 10)
```

raises an exception `AttributeError: 'int' object has no attribute
'get_sons'`.

Also:

```
f = (x == Int(10))
```

does not do what one would expect (i.e. it checks whether `x` and
`Int(10)` are equal, instead of building the equality formula).

**Comment by agriggio** [comment]
also, it would be nice to have infix notation for Boolean connectives &, |, ~

**Comment by marcogario** [comment]
Thanks for reporting.  In case it helps, you can use f.And() etc for
boolean infix operators.

**Comment by agriggio** [comment]
Yes, but it doesn't help :-) `f.And(f2)` is worse than `And(f, f2)`.
Infix notation is really handy in the shell, but I can live without it for more serious work.
However, automatic promotion of numbers would be nice to have (i.e. allow writing `Equals(x, 100)` instead of `Equals(x, Int(100)`)

**Comment by mikand** [comment]
There is a bit of ambiguity for ```x == 5``` it could be either ```x == Real(5)``` or ```x == Int(5)``` because PySMT distinguishes reals and integers.

This could be solved by forcing to use ```5.0``` for real numbers, but I'm a bit scared by numeric stability...

**Comment by mikand** [comment]
@marcogario: We decided to limit the infix notation for this and other issues (Boolean operators are not supported because of precedence issues and confusion between conjunction and bvand).
I'm for the "explicit is better than implicit", but since infix notation is optional we can go all-the-way with it... What do you think?

**Comment by agriggio** [comment]
Why is `x == 5` ambiguous? Isn't the type of `x` enough to resolve the ambiguity?
Regarding explicit vs implicit, I agree, but the infix notation is handy for playing around, but much less fun if it is only halfway... (Btw, if you really want to push it, `a & b` could mean both `And(a,b)` and  `BVand(a, b)`, depending on the argument types :-)

**Comment by mikand** [comment]
In cf8227f I gave a first shot at this issue:
I added &, |, ^, ~, -(unary) and the support for numeric constants. (Still missing BV overloads and casts)

The only operator I think is REALLY problematic is equality (==) as it is used (together with __hash__) when a formula is stored in a set or in a dict, so we cannot overload it....

**Comment by agriggio** [comment]
Thanks! 
regarding equality, how is `==` treated in Z3py?

**Comment by mikand** [comment]
In Z3py ```==``` is used to build an equality term as you suggested, but Z3 formulae cannot be used as elements of sets or maps.
Here is the hack suggested by Leonardo: http://stackoverflow.com/questions/12478429/hashing-expressions-in-z3python

**Comment by agriggio** [comment]
Hmmm... it might be a reasonable hack, I don't know. It is weird that
you have infix notation for everything but not for equality... One could
argue that defining `FNode.__hash__ = None` and have a separate wrapper
function `askey` agrees with the "explicit better than implicit" somehow
;-) Anyway, it's no big deal if you don't like it

**Comment by marcogario** [comment]
I would avoid overloading == . I understand that would make the use of the shell much nicer, but it would require changing all cases in which you want to compare nodes to an less intuitive way.

If we really think that infix notation is important for interactive usage, maybe we can take a closer look at something like : http://tomerfiliba.com/blog/Infix-Operators/

**Comment by agriggio** [comment]
hmmm, I don't know... since nodes are hashed, you can use "is" for comparison, no?
Also, you can make __nonzero__ raise an error to help people catch typos.
Anyway, now I'll stop arguing :-), as I said I can live without it

**Comment by marcogario** [comment]
Except for the == notation, the rest has been solved in PR #108.
For equality, for now, it is necessary to use FNode.Equals(). E.g.,
```
x.Equals(y)
```
